File: discord-collector/src/symbol-aliases.js
/**
 * 币安等所常见「千倍/百万倍」合约命名：
 * 1000PEPEUSDT → 展示/跨所查找用 PEPE；币安本所仍用 1000PEPEUSDT。
 */

const QUOTE_RE = /(?:USDT|USDC|BUSD|USD)$/i;
/** 较长前缀优先 */
const MULTIPLIER_PREFIX_RE = /^(1000000|100000|10000|1000|1M)/i;

/** @param {string} symbol */
export function stripQuote(symbol) {
  return String(symbol ?? "")
    .trim()
    .toUpperCase()
    .replace(QUOTE_RE, "");
}

/**
 * 去掉千倍前缀后的基础资产名：1000PEPE → PEPE，1MBABYDOGE → BABYDOGE
 * @param {string} symbol
 */
export function humanBaseAsset(symbol) {
  let base = stripQuote(symbol);
  if (!base) return "";
  const m = base.match(MULTIPLIER_PREFIX_RE);
  if (m && base.length > m[1].length) {
    base = base.slice(m[1].length);
  }
  return base;
}

/** @param {string} symbol @returns {string} 如 PEPEUSDT */
export function humanUsdtSymbol(symbol) {
  const base = humanBaseAsset(symbol);
  return base ? `${base}USDT` : "";
}

/** @param {string} symbol @returns {string} 规范为 *USDT */
export function toUsdtSymbol(symbol) {
  const s = String(symbol ?? "").trim().toUpperCase().replace(/[^A-Z0-9]/g, "");
  if (!s) return "";
  if (/USDT$|USDC$|BUSD$/.test(s)) return s.replace(/USDC$|BUSD$/, "USDT");
  return `${s}USDT`;
}

/**
 * 跨所 K 线/行情查找候选。
 * OKX/Bitget/Gate 常用人类名 PEPEUSDT；币安/Bybit 常用 1000PEPEUSDT。
 * 人类名会额外尝试 1000/1M 前缀（KORU → 1000KORUSDT）。
 * @param {string} symbol
 * @param {"binance"|"bybit"|"okx"|"bitget"|"gate"|string} [venue]
 */
export function symbolLookupCandidates(symbol, venue = "binance") {
  const raw = toUsdtSymbol(symbol);
  const human = humanUsdtSymbol(raw);
  const preferHuman = venue === "okx" || venue === "bitget" || venue === "gate";
  /** @type {string[]} */
  const multis = [];
  const base = humanBaseAsset(raw);
  const alreadyMulti = MULTIPLIER_PREFIX_RE.test(stripQuote(raw));
  if (base && !alreadyMulti) {
    for (const p of ["1000", "10000", "100000", "1000000", "1M"]) {
      multis.push(`${p}${base}USDT`);
    }
  }
  /** @type {string[]} */
  const ordered = preferHuman ? [human, raw, ...multis] : [raw, human, ...multis];
  /** @type {string[]} */
  const out = [];
  for (const s of ordered) {
    if (s && !out.includes(s)) out.push(s);
  }
  return out;
}

File: discord-collector/oi_mornitor/symbol_aliases.py
"""合约符号别名：币安千倍合约 ↔ 人类常用名。"""
from __future__ import annotations


def normalize_usdt_symbol(symbol: str) -> str:
    s = str(symbol or "").strip().upper().replace("-", "").replace("_", "")
    if not s:
        return ""
    if s.endswith(("USDC", "BUSD")):
        return f"{s[:-4]}USDT"
    if s.endswith("USDT"):
        return s
    return f"{s}USDT"


def human_base_asset(symbol: str) -> str:
    """1000PEPEUSDT → PEPE；币安千倍/百万倍合约的展示与跨所名。"""
    base = normalize_usdt_symbol(symbol)
    if base.endswith("USDT"):
        base = base[:-4]
    for prefix in ("1000000", "100000", "10000", "1000", "1M"):
        if base.startswith(prefix) and len(base) > len(prefix):
            return base[len(prefix) :]
    return base


def human_usdt_symbol(symbol: str) -> str:
    base = human_base_asset(symbol)
    return f"{base}USDT" if base else ""


_MULTIPLIER_PREFIXES = ("1000", "10000", "100000", "1000000", "1M")


def multiplier_usdt_symbols(symbol: str) -> list[str]:
    """人类名补千倍候选：KORU → 1000KORUSDT。"""
    base = human_base_asset(symbol)
    if not base:
        return []
    raw_base = normalize_usdt_symbol(symbol)
    if raw_base.endswith("USDT"):
        raw_base = raw_base[:-4]
    for prefix in _MULTIPLIER_PREFIXES:
        if raw_base.startswith(prefix) and len(raw_base) > len(prefix):
            return []
    return [f"{p}{base}USDT" for p in _MULTIPLIER_PREFIXES]


def symbol_lookup_candidates(symbol: str, venue: str = "binance") -> list[str]:
    """跨所候选。OKX/Bitget/Gate 优先 PEPEUSDT；币安/Bybit 再兜底 1000PEPEUSDT。"""
    raw = normalize_usdt_symbol(symbol)
    human = human_usdt_symbol(raw)
    multis = multiplier_usdt_symbols(raw)
    prefer_human = venue.lower() in ("okx", "bitget", "gate")
    ordered = [human, raw, *multis] if prefer_human else [raw, human, *multis]
    out: list[str] = []
    for s in ordered:
        if s and s not in out:
            out.append(s)
    return out


# 稳定币本位：几乎无波动，不进雷达/形态/回溯
STABLECOIN_BASES = frozenset({
    "USDT", "USDC", "FDUSD", "TUSD", "USDE", "USDD", "DAI", "BUSD",
    "USD1", "USDP", "GUSD", "FRAX", "PYUSD", "EURC", "EURT", "USTC",
})


def is_stablecoin_symbol(symbol: str) -> bool:
    """USDCUSDT / USDC / 1000USDCUSDT → True。"""
    s = str(symbol or "").strip().upper().replace("-", "").replace("_", "")
    if not s:
        return False
    if s in STABLECOIN_BASES:
        return True
    base = human_base_asset(s)
    return bool(base) and base.upper() in STABLECOIN_BASES
